doc/source/components/settings.rst
.. _settings:

Settings
--------

The Settings class provides a general-purpose data container for various kinds of information that need to be stored and processed by the pyZacros and PLAMS environment. It is formally identical that the Settings class in PLAMS. Please, see all details here `PLAMS.Settings <../../plams/components/settings.html>`_.

For our example (see :ref:`use case system <use_case_model_zgb>`), the following lines create the conditions we need for our calculation like e.g., the temperature (500 K) and pressure (1.0 atm.):

.. code-block:: python
  :linenos:

  # Settings:
  sett = pz.Settings()
  sett.random_seed = 953129
  sett.temperature = 500.0
  sett.pressure = 1.0
  sett.snapshots = ('time', 0.1)
  sett.process_statistics = ('time', 0.1)
  sett.species_numbers = ('time', 0.1)
  sett.event_report = 'off'
  sett.max_steps = 'infinity'
  sett.max_time = 1.0

  sett.molar_fraction.CO = 0.45
  sett.molar_fraction.O2 = 0.55

  print(sett)

As in the other pyZacros objects, the function ``print()`` (see line 16) shows the Settings object as it is going to be used
in the Zacros input files. The previous lines produce the following output:

.. code-block:: none

  random_seed         953129
  temperature          500.0
  pressure               1.0

  snapshots                 on time       0.1
  process_statistics        on time       0.1
  species_numbers           on time       0.1
  event_report      off
  max_steps         infinity
  max_time          1.0

Notice that the CO and O\ :sub:`2` molar fractions (keywords ``sett.molar_fraction.CO`` and ``sett.molar_fraction.O2``) are not printed out in the previous configuration block. This is because information about species involved in the mechanism and clusters is needed to generate the corresponding block in the ``simulation_input.dat`` zacros file. This information is going to print out from the :ref:`ZacrosJob <zacrosjob>` object.
Please consult Zacros' user guide (``$AMSHOME/scripting/scm/pyzacros/doc/ZacrosManual.pdf``) for more details about the specific meaning of the keywords used above.

API
~~~

.. currentmodule:: scm.pyzacros.core.Settings
.. autoclass:: Settings
    :exclude-members: __init__, __str__



core/Settings.py
"""
Module containing the Settings class.
"""

import scm.plams

__all__ = ['Settings']

class Settings( scm.plams.Settings ):
    """
    Automatic multi-level dictionary. Subclass of the PLAMS class `scm.plams.Settings <../../plams/components/settings.html>`_. This dictionary can contain any kind of information stored in key-value pairs. Be aware that no check of the rightness of the key/values is done at this level. This object is used just as a container. The verification of the physical meaning of the key-value pairs is done at the ZacrosJob class. The following is an example of use adapted to zacros:

    .. code:: python

        sett = Settings()
        sett.random_seed = 953129
        sett.temperature = 500.0
        sett.pressure = 1.0
        sett.snapshots = ('time', 0.1)
        sett.process_statistics = ('time', 0.1)
        sett.species_numbers = ('time', 0.1)
        sett.event_report = 'off'
        sett.max_steps = 'infinity'
        sett.max_time = 1.0
    """

    def __init__(self, *args, **kwargs):
        dict.__init__(self, *args, **kwargs)


    def __str__( self ):
        """
        Translates the object to a string
        """
        output  = ""

        if( 'random_seed' in self ): output += "random_seed     " + "%10s"%self.get('random_seed')+"\n"
        output += "temperature     " + "%10s"%self.get('temperature')+"\n"
        output += "pressure        " + "%10s"%self.get('pressure')+"\n\n"

        for option in ['snapshots', 'process_statistics', 'species_numbers']:
            if( option in self ):
                    pair = self[option]
                    if( len(pair) == 2 ):
                        key,value = pair
                    elif( len(pair) == 3 ):
                        key,value = pair[0],(pair[1],pair[2])
                    else:
                        msg  = "\n### ERROR ### keyword "+option+" in settings.\n"
                        msg += "              Its value should be a pair (key,value[,value1]).\n"
                        msg += "              Possible options for key:  'event', 'elemevent', 'time',       'logtime', 'realtime'\n"
                        msg += "              Possible options for value:  <int>,       <int>, <real>, (<real>,<real>),     <real>\n"
                        msg += "              Given value: "+str(pair)+"\n"
                        raise NameError(msg)

                    if( key == 'logtime' ):
                        if( len(value) != 2 ):
                            msg  = "\n### ERROR ### keyword '"+option+" on "+key+"' in settings.\n"
                            msg += "              Its value should be a pair of reals (<real>,<real>).\n"
                            msg += "              Given value: "+str(value)+"\n"
                            raise NameError(msg)

                        output += "%-20s"%option + "      " + "on "+ key + "       " + str(float(value[0])) + "  " + str(float(value[1])) + "\n"
                    elif( key == 'event' or key == 'elemevent' ):
                        output += "%-20s"%option + "      " + "on "+ key + "       " + str(int(value)) + "\n"
                    else:
                        output += "%-20s"%option + "      " + "on "+ key + "       " + str(float(value)) + "\n"

        if( 'event_report' in self ): output += "event_report      " + str(self.get(('event_report')))+"\n"
        if( 'max_steps' in self ): output += "max_steps         " + str(self.get(('max_steps')))+"\n"
        if( 'max_time' in self ): output += "max_time          " + str(self.get(('max_time')))+"\n"
        if( 'wall_time' in self ): output += "wall_time         " + str(self.get(('wall_time')))+"\n"

        if( 'override_array_bounds' in self ): output += "override_array_bounds         " + str(self.get(('override_array_bounds')))+"\n"

        return output
